File: moreparameters/moreparameters.js
"use strict";

// load moreparameters.node native module
const mylib = require("./build/Release/moreparameters.node");

var v_bool = true; // bool
var v_number = 15; // number
var v_array = [2,10,5]; // array

// print array
console.log("Array before: ", v_array);

// call with invalid count of parameter
console.log(mylib.moreparameters(v_bool));

// print changed array
console.log("Array after: ", mylib.moreparameters(v_bool, v_number, v_array));

File: moreparameters/moreparameters.cpp
#include <napi.h>

using namespace Napi;

// function moreparameters
Value moreparameters(const CallbackInfo &info)
{
    // check if parameter count is valid
    if (info.Length() != 3)
        return String::New(info.Env(), "Invalid call");

    // get parameters
    bool v_bool = info[0].As<Boolean>();
    uint32_t v_number = info[1].As<Number>().Uint32Value();
    Array v_array = info[2].As<Array>();

    // change array elements
    v_array[uint32_t(0)] = Number::New(info.Env(), v_number);
    v_array[uint32_t(1)] = String::New(info.Env(), "String element");
    v_array[uint32_t(2)] = Boolean::New(info.Env(), !v_bool);
    
    // return changed array
    return v_array;
}

// function adds symbols to export list
Object export_all(Env env, Object exports)
{
    exports.Set("moreparameters", Function::New(env, moreparameters));
    return exports;
}

// macro exports all symbols
NODE_API_MODULE(NODE_GYP_MODULE_NAME, export_all);


